### ISSUE TITLE
overloaded function body is unsafe, it should be more strict

### ISSUE THREAD

**Issue by safareli** [microsoft/TypeScript#43751]
# Bug Report

### 🔎 Search Terms

overload unsafe function

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about overloading

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.3.3&ssl=1&ssc=1&pln=9&pc=1#code/GYVwdgxgLglg9mABMMAKAHgLkQZygJxjAHMBKbPQkgbgChRJYFk0tEwQBbAIwFN9yibnDgAbXgEMwdBtHhIUGCgSLFEAH3Zc+A5VTWbhYyUgDetRIny8oIfEgBEIgA44HdAL61aAeh+J0RGwjcSlaCAQ8AMQAXhZUAEZSOiA)

### 💻 Code

```ts
function fn(x: string): string;
function fn(x: number): boolean;
function fn(x: string | number): string | boolean {
  return "oops";
}

// x : boolean
const x = fn(1);
// in runtime `typeof x` will be `"string"`
```

### 🙁 Actual behavior

Currently typescript allows an unsafe implementation to be provided in overloaded function body.

### 🙂 Expected behavior

it's expected for typescript to understand in overloaded function body that  if `x` is number only then should boolean be returned or if `x` is string only then should string  be returned. i.e. I should get error   on `return "oops";`

**Comment by RyanCavanaugh** [comment]
Duplicate #13258 and others. We don't intend to implement this sort of check.

**Comment by SRachamim** [comment]
I'm shocked to find out that type-safety is not a concern of _TypeScript_. At least leave it open on the backlock instead of closing it and limiting the discussion about it.

**Comment by RyanCavanaugh** [comment]
Regarding leaving issues open, the "closed" bit is just a color on a webpage and should be treated as such. You can feel however you want about it, but we require the ability to make decisions and indicate the results of those decisions. The decision here is "We're not doing this anytime soon". We've communicated that decision by closing the issue. This doesn't mean you can't comment here, or point out areas where the feature would deliver more value than we thought it would, or engage in other ways. The decision has been made for now not to do this and it'd be a disservice to anyone coming across the issue to think that it was on the table for the near future.

**Comment by SRachamim** [comment]
> @microsoft microsoft locked and limited conversation to collaborators on Jun 20, 2018
Doesn't seem like I can join in two years after and throw my two cents.

**Comment by RyanCavanaugh** [comment]
Your comments are showing up here and someone is responding to you. ?

**Comment by SRachamim** [comment]
Here, not in the original issue. But possible not for long here too.

**Comment by SRachamim** [comment]
> Regarding leaving issues open, the "closed" bit is just a color on a webpage and should be treated as such.
It's not a colour. It's a statistic. When I use a library I want to see how many unclosed issues it has. When you close an issue that wasn't actually closed - then you hide it from the statistics.

If you agree this is a bug, and it should be fixed "someday" ideally - then it's an open issue that should not be closed.

The colour can be yellow, red, or green - but the numbers should reflect reality.

**Comment by RyanCavanaugh** [comment]
>  When I use a library I want to see how many unclosed issues it has.

This is precisely why we close things we don't intend to do. It would be silly to leave open *every conceivable improvement* to the language -- that list is literally infinite. Open/closed status should communicate something.

> If you agree this is a bug, and it should be fixed "someday" ideally - then it's an open issue that should not be closed.

It's a suggestion, and I don't agree that this is the best use of complexity budget compared to the other options.

> The colour can be yellow, red, or green

GitHub only has two issue states: Open and closed.

We have our definition of open/closed here. I understand you might like a different definition, but this is a choice of the maintainers. Everyone has their personal definition of what open/closed should mean, and if we entertained debate on the matter, it would be literally all we did. I don't intend to discuss this further.
